components/builder-api-client/src/lib.rs

```diff
@@ -199,22 +201,84 @@ pub struct OriginChannelIdent {
 
 #[derive(Clone, Deserialize)]
 pub struct OriginInvitation {
-    pub id:           String,
-    pub account_id:   String,
+    #[serde(with = "json_u64")]
+    pub id: u64,
+    #[serde(with = "json_u64")]
+    pub account_id: u64,
     pub account_name: String,
-    pub created_at:   String,
-    pub ignored:      bool,
-    pub origin:       String,
-    pub owner_id:     String,
-    pub updated_at:   String,
+    pub created_at: String,
+    pub ignored: bool,
+    pub origin: String,
+    #[serde(with = "json_u64")]
+    pub owner_id: u64,
+    pub updated_at: String,
 }
 
+#[derive(Clone, Deserialize)]
+pub struct UserOriginInvitationsResponse(pub Vec<OriginInvitation>);
+
 #[derive(Clone, Deserialize)]
 pub struct PendingOriginInvitationsResponse {
     pub origin:      String,
     pub invitations: Vec<OriginInvitation>,
 }
 
+macro_rules! invite_format {
+    () => {
+        "{:>20} | {:>20} | {:>20} | {:>20} | {:>10}"
+    };
+}
+
+macro_rules! invite_header {
+    () => {
+        format!(invite_format!(),
+                "Id", "Origin", "Account", "Date", "Ignored")
+    };
+}
+
+impl fmt::Display for OriginInvitation {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        let mut output = Vec::new();
+        let created_at =
+            match NaiveDateTime::parse_from_str(&self.created_at, "%Y-%m-%dT%H:%M:%S%.f") {
+                Ok(date) => format!("{}", date.format("%Y-%m-%d %H:%M:%S")),
+                Err(_) => format!("{}", Utc::now()),
+            };
+        output.push(format!(invite_format!(),
+                            self.id, self.origin, self.account_name, created_at, self.ignored));
+        write!(f, "{}", output.join("\n"))
```

Comment:
I have some misgivings about removing `TabWriter` completely in favor of custom `Display` implementations on these types.

From my perspective, this is taking a very specific output need (i.e., creating a table) and elevating it to the default output rendering. While you _might_ be able to make an argument for that from the point of view of the various response structs below, I think it's more difficult to do that here. For instance, the printing of an `OriginInvitation` is now implicitly assumed to be taking place _inside a table_, to the point that you're also embedding a newline at the end of formatted output (the embedded newline by itself is a bit of a red flag to me, actually).

While this is indeed the only way that we're outputting an `OriginInvitation` now, there's nothing to say that's the way it will be forever. As soon as another output need is uncovered, it will become rather awkward to work with this.

The previous way of doing this with TabWriter was OK, in that it was basically encoding a new, situation-specific rendering behavior for a group of items; I think it was just repeated more than necessary.

If you wanted to go all the way down the rabbit hole, you could envision some kind of `Tabular` trait that you could implement that would allow you to define a `header()` and `as_table_row(&self)` function for a given type. Then when you're dealing with collections of a `Tabular` implementation, you could generically call the `header()` function first, then iterate through the collection and call `as_table_row` for each item. The nice thing about this is that you can encapsulate all that behavior in a single place, and then easily get tabular output for anything by implementing two simple functions. The generic implementation could remain free to use `TabWriter` as an implementation detail.

That makes it very straightforward to see what kind of behavior is being used in each situation, and doesn't assume too much in the general case.

(You don't necessarily need to do all that here, but it could be used as a foundation on which to squeeze more standardization out of our CLI in the future.)

Reply:
As an aside, the use of a `Vec` in this implementation is unnecessary, since you're only ever creating a single string. You could just as easily add the newline into the `write!` formatting string and get the same outcome (or, probably, just use `writeln!`).

(However, I think embedding the newline in a `Display` implementation is a bit suspect to begin with!)

Reply:
I'll follow up on your suggestions.

Reply:
@christophermaier I've introduced a Tabular trait in https://github.com/habitat-sh/habitat/pull/7349/commits/a65aebfafec7355e188e527768485ba9957fa404

Since this thread references `impl fmt::Display for OriginInvitation` which is no longer in the code base (the implementation has been moved to a trait), I'm going to mark this resolved and ask for a fresh review 😄